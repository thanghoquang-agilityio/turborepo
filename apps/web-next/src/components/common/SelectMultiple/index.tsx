'use client'

import { ChangeEvent, memo, useCallback, useMemo, useState } from 'react'

// Component
import { Select, SelectItem } from '@nextui-org/react'

// Hooks
import { useDebounce } from '@/hooks'
// Types
import { FilterItem } from '@/types'

interface SelectMultipleProps {
  items: FilterItem[]
  placeholder?: string
  defaultValue?: string[]
  additionalClass?: string
  isDisabled?: boolean
  handleReplaceURL?: (params: URLSearchParams) => void
}

const SelectMultiple = memo(
  ({
    items,
    placeholder = '-- Select --',
    defaultValue = [],
    additionalClass = '',
    isDisabled = false,
    handleReplaceURL,
  }: SelectMultipleProps) => {
    const [values, setValues] = useState(new Set(defaultValue))
    const searchParams = useMemo(() => new URLSearchParams(), [])

    const updateSearchParams = useCallback(
      (values: string[], items: { [key: number]: { value: string } }) => {
        searchParams.delete('category')

        values
          .filter((element) => element !== '')
          .forEach((value) => {
            const valueSplit = value.split('-')
            const index = Number(valueSplit[1])

            const itemValue = items[index].value

            if (!searchParams.getAll('category').includes(itemValue)) {
              searchParams.append('category', itemValue)
            }
          })
        handleReplaceURL?.(searchParams)
      },
      [handleReplaceURL, searchParams]
    )

    const debounceSelect = useDebounce((value: Set<string>) => {
      if (!value) return

      updateSearchParams(Array.from(value), items)
    })

    const handleSelectionChange = useCallback(
      (e: ChangeEvent<HTMLSelectElement>) => {
        setValues(new Set(e.target.value.split(',')))
        debounceSelect(new Set(e.target.value.split(',')))
      },
      [debounceSelect]
    )

    return (
      <Select
        className={`bg-background-400 text-sm font-medium text-primary-200 ${additionalClass}`}
        classNames={{
          base: 'rounded-[28px] opacity-100',
          trigger: 'min-h-9 h-9 px-3.5 rounded-[28px]',
          value: 'text-primary-200 font-medium text-sm',
        }}
        selectionMode='multiple'
        placeholder={placeholder}
        selectedKeys={values}
        onChange={handleSelectionChange}
        isDisabled={isDisabled}
        aria-label={placeholder}
        data-testid={placeholder}
      >
        {items.map((item: FilterItem, index) => (
          <SelectItem
            key={`${placeholder}-${index}`}
            value={item.value}
            className='data-[hover=true]:text-text-primary-200 rounded-[28px] text-sm font-medium data-[selectable=true]:focus:bg-background-200 data-[selectable=true]:focus:text-primary-200'
          >
            {item.name}
          </SelectItem>
        ))}
      </Select>
    )
  }
)

SelectMultiple.displayName = 'SelectMultiple'
export default SelectMultiple
