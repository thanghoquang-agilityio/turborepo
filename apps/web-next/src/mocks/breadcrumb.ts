export const MOCK_BREADCRUMB_ITEMS = [
  { name: 'Products', link: 'products' },
  { name: 'Detail Product', link: '7' },
]
