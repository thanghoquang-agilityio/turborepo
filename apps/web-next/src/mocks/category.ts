export const MOCK_CATEGORY_LIST = [
  {
    name: 'Kid',
    value: 'kid',
    total: 18,
  },
  {
    name: 'Man',
    value: 'man',
    total: 12,
  },
  {
    name: 'Woman',
    value: 'woman',
    total: 23,
  },
  {
    name: 'Sport',
    value: 'sport',
    total: 34,
  },
]

export const MOCK_CATEGORY_RESPONSE = [
  {
    name: 'Kid',
    value: 'kid',
    products: {
      data: [],
    },
  },
  {
    name: 'Man',
    value: 'man',
    products: {
      data: [],
    },
  },
  {
    name: 'Woman',
    value: 'woman',
    products: {
      data: [],
    },
  },
  {
    name: 'Sport',
    value: 'sport',
    products: {
      data: [],
    },
  },
]
