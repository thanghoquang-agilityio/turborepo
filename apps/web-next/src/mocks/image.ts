import { SRC_IMAGE_NOT_AVAILABLE } from '@/constants'

export const MOCK_IMAGE = {
  data: {
    id: '24',
    attributes: {
      name: 'default',
      url: SRC_IMAGE_NOT_AVAILABLE,
    },
  },
}

export const MOCK_LIST_IMAGE_RESPONSE = [
  {
    name: 'default1',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
  {
    name: 'default2',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
  {
    name: 'default3',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
  {
    name: 'default4',
    url: SRC_IMAGE_NOT_AVAILABLE,
  },
]
