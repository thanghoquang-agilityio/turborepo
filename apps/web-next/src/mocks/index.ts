export * from './breadcrumb'
export * from './cart'
export * from './category'
export * from './image'
export * from './product'
export * from './radio-group'
export * from './user'
