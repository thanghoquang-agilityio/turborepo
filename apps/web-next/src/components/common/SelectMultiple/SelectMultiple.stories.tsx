import type { Meta, StoryObj } from '@storybook/react'

import SelectMultiple from '@/components/common/SelectMultiple'
// Mocks
import { MOCK_CATEGORY_LIST } from '@/mocks'

const meta = {
  title: 'Components/SelectMultiple',
  tags: ['autodocs'],
  component: SelectMultiple,
  parameters: {
    layout: 'centered',
    nextjs: {
      appDirectory: true,
    },
  },
} satisfies Meta<typeof SelectMultiple>

export default meta
type Story = StoryObj<typeof SelectMultiple>

export const Default: Story = {
  args: {
    items: MOCK_CATEGORY_LIST,
    additionalClass: 'w-[200px]',
  },
}
