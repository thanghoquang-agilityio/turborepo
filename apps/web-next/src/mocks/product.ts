import { ProductVariantModel } from '@/types'

import { MOCK_IMAGE } from './image'

// New flattened product variants for testing
export const MOCK_PRODUCT_VARIANTS: ProductVariantModel[] = [
  {
    id: 1,
    color: 'slate blue',
    size: 'small,medium,large,extra large,XXL',
    stock: 20,
    tax: 5,
  },
  {
    id: 2,
    color: 'blue',
    size: 'small,medium,large,extra large,XXL',
    stock: 20,
    tax: 5,
  },
]

export const MOCK_PRODUCT = {
  id: '1',
  image: MOCK_IMAGE,
  name: 'TDX Sinkers',
  variants: 5,
  price: 675.0,
  star: 5,
  comments: 121,
  description: '',
  discount: 0,
  likes: 11,
  brand: 'MOCK BRAND',
  productVariants: {
    data: MOCK_PRODUCT_VARIANTS,
  },
}

export const MOCK_PRODUCTS_RESPONSE = [
  {
    id: '1',
    attributes: {
      name: 'TDX Sinkers',
      variants: 5,
      price: 675.0,
      star: 5,
      comments: 121,
      brand: 'MOCK BRAND',
      description: '',
      discount: 0,
      likes: 11,
      productVariants: {
        data: [],
      },
    },
  },
  {
    id: '2',
    attributes: {
      name: 'TDX Sinkers',
      variants: 5,
      price: 675.0,
      star: 5,
      comments: 121,
      brand: 'MOCK BRAND',
      description: '',
      discount: 0,
      likes: 11,
      productVariants: {
        data: [],
      },
    },
  },
  {
    id: '3',
    attributes: {
      name: 'TDX Sinkers',
      variants: 5,
      price: 675.0,
      star: 5,
      comments: 121,
      brand: 'MOCK BRAND',
      description: '',
      discount: 0,
      likes: 11,
      productVariants: {
        data: [],
      },
    },
  },
  {
    id: '4',
    attributes: {
      name: 'TDX Sinkers',
      variants: 5,
      price: 675.0,
      star: 5,
      comments: 121,
      brand: 'MOCK BRAND',
      description: '',
      discount: 0,
      likes: 11,
      productVariants: {
        data: MOCK_PRODUCT_VARIANTS,
      },
      image: MOCK_IMAGE,
    },
  },
]
