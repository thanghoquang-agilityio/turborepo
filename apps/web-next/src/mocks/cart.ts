import { CartItemResponse, LegacyCartItemResponse } from '@/types'
import { MOCK_PRODUCT_VARIANTS } from './product'

export const MOCK_CART_ITEM = {
  id: '1',
  variantId: '1',
  name: 'Embrace Sideboard',
  color: 'wheat',
  size: 'small',
  quantityDefault: 2,
  stock: 4,
  price: 53,
  star: 1,
  comments: 1,
  tax: 1,
  discount: 1,
}

// New flattened cart response structure
export const MOCK_CART_RESPONSE: CartItemResponse[] = [
  {
    id: 1,
    documentId: 'cart-item-1',
    size: 'small',
    quantity: 2,
    createdAt: '2024-01-01T00:00:00.000Z',
    updatedAt: '2024-01-01T00:00:00.000Z',
    publishedAt: '2024-01-01T00:00:00.000Z',
    productVariantId: MOCK_PRODUCT_VARIANTS[0],
    userId: 1,
  },
  {
    id: 2,
    documentId: 'cart-item-2',
    size: 'large',
    quantity: 2,
    createdAt: '2024-01-01T00:00:00.000Z',
    updatedAt: '2024-01-01T00:00:00.000Z',
    publishedAt: '2024-01-01T00:00:00.000Z',
    productVariantId: MOCK_PRODUCT_VARIANTS[1],
    userId: 1,
  },
]

// Legacy cart response for backward compatibility
export const MOCK_CART_RESPONSE_LEGACY: LegacyCartItemResponse[] = [
  {
    id: '1',
    attributes: {
      productVariantId: {
        data: {
          id: '1',
          attributes: {
            size: 'small,medium,large,extra large,XXL',
            color: 'slate blue',
            stock: '2,10,0,23,20',
            tax: 12,
          },
        },
      },
      size: 'small',
      quantity: 2,
    },
  },
  {
    id: '2',
    attributes: {
      productVariantId: {
        data: {
          id: '1',
          attributes: {
            size: 'small,medium,large,extra large,XXL',
            color: 'slate blue',
            stock: '20,10,20,23,20',
            tax: 12,
          },
        },
      },
      size: 'large',
      quantity: 2,
    },
  },
]

export const MOCK_ORDER_SUMMARY = {
  subTotal: 396,
  itemTotal: 3,
}
