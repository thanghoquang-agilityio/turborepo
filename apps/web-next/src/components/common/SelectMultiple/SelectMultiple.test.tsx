import { fireEvent, render, screen, waitFor } from '@testing-library/react'

import { MOCK_CATEGORY_LIST } from '@/mocks'

import SelectMultiple from '.'

describe('SelectMultiple Component', () => {
  const renderComponent = (props = {}) => {
    return render(<SelectMultiple items={MOCK_CATEGORY_LIST} {...props} />)
  }

  it('renders with default props', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('handles selection change', () => {
    const { getByTestId } = renderComponent()
    const selectElement = getByTestId('-- Select --')

    fireEvent.change(selectElement, { target: { value: 'Man' } })

    waitFor(() => {
      expect(screen.getByText('Man')).toBeInTheDocument()
    })
  })
})
