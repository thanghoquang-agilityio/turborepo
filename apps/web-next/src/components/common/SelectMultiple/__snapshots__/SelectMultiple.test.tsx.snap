// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectMultiple Component renders with default props 1`] = `
<div>
  <div
    class="group flex flex-col w-full transition-background motion-reduce:transition-none !duration-150 group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)] rounded-[28px] opacity-100 bg-background-400 text-primary-200 font-medium text-sm"
    data-filled="true"
    data-slot="base"
  >
    <div
      aria-hidden="true"
      data-a11y-ignore="aria-hidden-focus"
      data-testid="hidden-select-container"
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        style="font-size: 16px;"
        tabindex="-1"
        type="text"
        value=""
      />
      <label>
        <select
          multiple=""
          size="4"
          tabindex="-1"
          title=""
        >
          <option />
          <option
            value="-- Select ---0"
          >
            Kid
          </option>
          <option
            value="-- Select ---1"
          >
            Man
          </option>
          <option
            value="-- Select ---2"
          >
            Woman
          </option>
          <option
            value="-- Select ---3"
          >
            Sport
          </option>
        </select>
      </label>
    </div>
    <div
      class="w-full flex flex-col"
      data-slot="mainWrapper"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="-- Select --"
        aria-labelledby="react-aria-:r7: react-aria-:r2:"
        class="relative gap-3 w-full inline-flex flex-row items-center shadow-sm tap-highlight-transparent bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 min-h-9 h-9 px-3.5 rounded-[28px]"
        data-slot="trigger"
        data-testid="-- Select --"
        id="react-aria-:r2:"
        type="button"
      >
        <div
          class="inline-flex h-full w-[calc(100%_-_theme(spacing.6))] min-h-4 items-center gap-1.5 box-border"
          data-slot="innerWrapper"
        >
          <span
            class="w-full text-left rtl:text-right truncate group-data-[has-value=true]:text-default-foreground text-primary-200 font-medium text-sm"
            data-slot="value"
            id="react-aria-:r7:"
          >
            -- Select --
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="absolute right-3 rtl:left-3 rtl:right-[unset] w-4 h-4 transition-transform duration-150 ease motion-reduce:transition-none data-[open=true]:rotate-180"
          data-slot="selectorIcon"
          fill="none"
          focusable="false"
          height="1em"
          role="presentation"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
